java8/proxy/src/main/java/openwhisk/java/action/Proxy.java

```diff
@@ -14,156 +14,198 @@
+          String clazzName = value.getString("main");
+          String methodName = "main";
+          if(clazzName == null || clazzName.isEmpty()){
+            clazzName = getClassNameFromManifest(url);
+          }
+          if(clazzName.indexOf('#')>-1){
+            final String[] splitted = clazzName.split("#");
+            clazzName = splitted[0];
+            methodName= splitted[1];
+          }
+          if(clazzName == null ){
+            rc.response().setStatusCode(ERROR_FUNCTION_NOT_FOUND).end("Main class is empty, can not determine the function to invoke");
+            return;
+          }
+          final Class<?> mainClass = urlClassLoader.loadClass(clazzName);
+
+          final Method m = mainClass.getMethod(methodName, new Class[] { com.google.gson.JsonObject.class });
+          m.setAccessible(true);
+          final int modifiers = m.getModifiers();
+          if (m.getReturnType() != com.google.gson.JsonObject.class || !Modifier.isStatic(modifiers) || !Modifier.isPublic(modifiers)) {
+            throw new NoSuchMethodException(methodName);
+          }
+          this.mainMethod = m;
+
+          rc.response().setStatusCode(HTTP_OK).end("OK");
+
+        }catch(NoSuchMethodException e){
+          rc.response().setStatusCode(ERROR_FUNCTION_NOT_FOUND).end(e);
+        }
+        catch (ClassNotFoundException | IOException e) {
+          rc.response().setStatusCode(HTTP_INTERNAL_SERVER_ERROR).end("An error has occurred (see logs for details): " + e);
+        }
+      }
+    });
+  }
+
+  private String getClassNameFromManifest(URL url ) throws IOException{
+    final URL jarUrl = new URL("jar", "", url + "!/");
+    final JarURLConnection connection = (JarURLConnection)jarUrl.openConnection();
+    final Attributes attributes = connection.getMainAttributes();
+    return attributes != null
+                   ? attributes.getValue(Attributes.Name.MAIN_CLASS)
+                   : null;
+  }
+
+  private void runHandler(RoutingContext rc) {
+    final JsonObject request = rc.getBodyAsJson();
+    final JsonObject value = request.getJsonObject("value");
+    final JsonParser parser = new JsonParser();
+    final com.google.gson.JsonObject req = parser.parse(value.toString()).getAsJsonObject();
+
+    final HashMap<String, String> env = new HashMap<>();
+    for (String envKey : OW_ENV_KEYS) {
+      env.put(String.format("__OW_%s", envKey.toUpperCase()), value.getString(envKey));
     }
-
-    private static void writeError(HttpExchange t, String errorMessage) throws IOException {
-        JsonObject message = new JsonObject();
-        message.addProperty("error", errorMessage);
-        writeResponse(t, 502, message.toString());
+    augmentEnv(env);
+
+    Thread.currentThread().setContextClassLoader(urlClassLoader);
+    System.setSecurityManager(new WhiskSecurityManager());
+
+    try {
+      final com.google.gson.JsonObject out = (com.google.gson.JsonObject) this.mainMethod.invoke(null, req);
```

Comment:
i think if you leveraged the vertx features like eventbus the function invocation would be much more reactive.
But, i think new action "kind" would be necessary as the method invocation should make the things a different way,.... 
what if you invoked function using event bus, e.g.:
```
   vertx.eventBus().<JsonObject>send("activation", <inputJsonObject>, reply - > {
       if (reply.failed()) {
          rc.response().setStatusCode(500);
       } else {
         rc.response.end(...);
      }
  }
```

Then the function could be just a class extending `AbstractVerticle` and in `start()` method simply consume eventBus instead of class with `main` method.